src/python/pants/backend/project_info/peek.py

```diff
@@ -106,6 +107,8 @@ def default(self, o):
             return list(o)
         if isinstance(o, Field):
             return self.default(o.value)
+        if isinstance(o, PythonArtifact):
```

Comment:
I'm wondering if we should implement this as runtime-checkable protocol rather than asking for a concrete type here:

```python
from typing import Any, Mapping, Protocol, runtime_checkable

@runtime_checkable
class Dictable(Protocol):
    def asdict(self) -> Mapping[str, Any]: ...


# then here the isinstance check is just this:
if isinstance(o, Dictable):
    return o.asdict()
```

Maybe the details aren't perfect, but I think this avoids us getting in a situation where we need to keep adding concrete types to this method to serialize things. Stuff that wasn't to be serializable here just needs to implement Dictable. They wouldn't even need to import it, just implement the `asdict(self)` method and it works.

Reply:
Oh that's juicy, I love it. Let me experiment locally first, but this is perfect!

Reply:
I had a similar comment to make, but without the concrete suggestion, so thanks for that :D